common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationFactory.java

```diff
@@ -828,4 +843,52 @@ public static byte[] getCollationKeyBytes(UTF8String input, int collationId) {
     }
   }
 
+  /**
+   * Returns same string if collation name is valid or the closest suggestion if it is invalid.
+   */
+  public static String getClosestSuggestionOnInvalidName(String collationName) {
+    String[] validRootNames;
+    String[] validModifiers;
+    if (collationName.startsWith("UTF8_")) {
+      validRootNames = new String[]{
+        Collation.CollationSpecUTF8Binary.UTF8_BINARY_COLLATION.collationName,
+        Collation.CollationSpecUTF8Binary.UTF8_LCASE_COLLATION.collationName
+      };
+      validModifiers = new String[0];
+    } else {
+      validRootNames = getICULocaleNames();
+      validModifiers = new String[]{"_CI", "_AI", "_CS", "_AS"};
+    }
+
+    // Split modifiers and locale name.
+    final int MODIFIER_LENGTH = 3;
+    String localeName = collationName.toUpperCase();
+    List<String> modifiers = new ArrayList<>();
+    while (Arrays.stream(validModifiers).anyMatch(localeName::endsWith)) {
+      modifiers.add(localeName.substring(localeName.length() - MODIFIER_LENGTH));
+      localeName = localeName.substring(0, localeName.length() - MODIFIER_LENGTH);
+    }
+
+    // Suggest version with unique modifiers.
+    Collections.reverse(modifiers);
+    modifiers = modifiers.stream().distinct().toList();
+
+    // Remove conflicting settings.
+    if (modifiers.contains("_CI") && modifiers.contains(("_CS"))) {
+      modifiers = modifiers.stream().filter(m -> !m.equals("_CI")).toList();
+    }
+
+    if (modifiers.contains("_AI") && modifiers.contains(("_AS"))) {
+      modifiers = modifiers.stream().filter(m -> !m.equals("_AI")).toList();
+    }
+
+    // Find the closest locale name.
+    final String finalLocaleName = localeName;
+    String closestLocale = Collections.min(List.of(validRootNames), Comparator.comparingInt(
```

Comment:
shall we return more than one proposals?

Reply:
FYI, for column not exist error, we provide up to 5 proposals. It may be too much for string collations, but sometimes the distances are very close and it makes more sense to return more than one proposals.

Reply:
I initially went with 3 proposals but result looked weird - locale names are sometimes pretty distant so suggestions were very far off.

In the latest iteration I propose following:
1) Take top 3.
2) Always return the closest one.
3) For the other 2, check if their distance is < levenshtein distance threshold (number of characters needed to be changed in order to reach the correct name). If we are below the limit we include other suggestions as well.

Btw, soon we will include some way to list all collations (e.g. `SHOW COLLATIONS`) and extend error message to point to a way how to see all collations available on the system.